File: threads/src/main/java/im/threads/internal/adapters/FilesAndMediaAdapter.java
package im.threads.internal.adapters;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import im.threads.internal.holders.EmptyViewHolder;
import im.threads.internal.holders.FileAndMediaViewHolder;
import im.threads.internal.holders.FilesDateStampHolder;
import im.threads.internal.model.DateRow;
import im.threads.internal.model.FileAndMediaItem;
import im.threads.internal.model.FileDescription;
import im.threads.internal.model.MediaAndFileItem;
import im.threads.internal.utils.FileUtils;

/**
 * Created by yuri on 01.07.2016.
 */
public final class FilesAndMediaAdapter extends RecyclerView.Adapter {
    private static final int TYPE_DATE_ROW = 1;
    private static final int TYPE_FILE_AND_MEDIA_ROW = 2;
    private OnFileClick mOnFileClick;

    private ArrayList<MediaAndFileItem> list;
    private ArrayList<MediaAndFileItem> backup;

    public FilesAndMediaAdapter(List<FileDescription> list, OnFileClick onFileClick) {
        this.list = new ArrayList<>();
        mOnFileClick = onFileClick;
        if (list.size() != 0) {
            addItems(list);
        }
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case TYPE_DATE_ROW:
                return new FilesDateStampHolder(parent);
            case TYPE_FILE_AND_MEDIA_ROW:
                return new FileAndMediaViewHolder(parent);
            default:
                return new EmptyViewHolder(parent);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, final int position) {
        if (getItemViewType(position) == TYPE_DATE_ROW) {
            FilesDateStampHolder h = (FilesDateStampHolder) holder;
            h.onBind(list.get(position).getTimeStamp());
        }
        if (getItemViewType(position) == TYPE_FILE_AND_MEDIA_ROW) {
            final FileAndMediaViewHolder h = (FileAndMediaViewHolder) holder;
            FileAndMediaItem item = (FileAndMediaItem) list.get(position);
            h.onBind(
                    item.getFileDescription(),
                    v -> {
                        if (null != mOnFileClick) {
                            FileAndMediaItem item1 = (FileAndMediaItem) list.get(h.getAdapterPosition());
                            mOnFileClick.onFileClick(item1.getFileDescription());
                        }
                    }
            );
        }
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (list.get(position) instanceof DateRow) {
            return TYPE_DATE_ROW;
        }
        if (list.get(position) instanceof FileAndMediaItem) {
            return TYPE_FILE_AND_MEDIA_ROW;
        }
        return super.getItemViewType(position);
    }

    public void backupAndClear() {
        backup = new ArrayList<>(list);
        list.clear();
    }

    public void filter(String filter) {
        if (filter == null) filter = "";
        list.clear();
        ArrayList<FileDescription> filteredItems = new ArrayList<>();
        for (MediaAndFileItem maf : backup) {
            if (maf instanceof FileAndMediaItem) {
                FileAndMediaItem fileAndMediaItem = (FileAndMediaItem) maf;
                FileDescription fd = fileAndMediaItem.getFileDescription();
                if (fd != null) {
                    String lastPathSegment = fd.getFileUri() != null ? fd.getFileUri().getLastPathSegment() : null;
                    if (lastPathSegment != null && lastPathSegment.toLowerCase().contains(filter.toLowerCase())) {
                        filteredItems.add(fd);
                    } else if (fd.getIncomingName() != null) {
                        String name = fd.getIncomingName();
                        if (name.toLowerCase().contains(filter.toLowerCase())) {
                            filteredItems.add(fd);
                        }
                    } else if (fileAndMediaItem.getFileName().toLowerCase().contains(filter.toLowerCase())) {
                        filteredItems.add(fd);
                    }
                }
            }
        }
        addItems(filteredItems);
        notifyDataSetChanged();
    }

    public void undoClear() {
        list = new ArrayList<>(backup);
        backup.clear();
        notifyDataSetChanged();
    }

    private void addItems(List<FileDescription> fileDescriptionList) {
        if (fileDescriptionList.size() == 0) return;
        Calendar current = Calendar.getInstance();
        Calendar prev = Calendar.getInstance();
        if (list.size() == 0) {
            FileDescription fd = fileDescriptionList.get(0);
            this.list.add(new DateRow(fd.getTimeStamp()));
            this.list.add(new FileAndMediaItem(fd, fd.getFileUri() != null ? FileUtils.getFileName(fd.getFileUri()) : ""));

        }
        for (int i = 1; i < fileDescriptionList.size(); i++) {
            FileDescription fd = fileDescriptionList.get(i);
            current.setTimeInMillis(fd.getTimeStamp());
            prev.setTimeInMillis(fileDescriptionList.get(i - 1).getTimeStamp());
            this.list.add(new FileAndMediaItem(fd, fd.getFileUri() != null ? FileUtils.getFileName(fd.getFileUri()) : ""));
            if (current.get(Calendar.DAY_OF_YEAR) != prev.get(Calendar.DAY_OF_YEAR)) {
                this.list.add(new DateRow(fd.getTimeStamp()));
            }
        }
    }

    public interface OnFileClick {
        void onFileClick(FileDescription fileDescription);
    }
}

File: threads/src/main/java/im/threads/internal/adapters/files_and_media/FilesAndMediaAdapter.kt
package im.threads.internal.adapters.files_and_media

import android.view.ViewGroup
import androidx.core.util.ObjectsCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import im.threads.internal.holders.EmptyViewHolder
import im.threads.internal.holders.FileAndMediaViewHolder
import im.threads.internal.holders.FilesDateStampHolder
import im.threads.internal.model.DateRow
import im.threads.internal.model.FileAndMediaItem
import im.threads.internal.model.FileDescription
import im.threads.internal.model.MediaAndFileItem
import im.threads.internal.utils.FileUtils.getFileName
import java.util.Calendar
import java.util.Locale

/**
 * Адаптер для списка файлов, отправленных или полученных в чате
 */
@Suppress("NAME_SHADOWING")
internal class FilesAndMediaAdapter(
    filesList: List<FileDescription?>,
    private val onFileClick: OnFileClick,
    private val onDownloadFileClick: OnFileClick
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var list = ArrayList<MediaAndFileItem>()
    private var backup = ArrayList<MediaAndFileItem>()

    init {
        addItems(filesList)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            TYPE_DATE_ROW -> FilesDateStampHolder(parent)
            TYPE_FILE_AND_MEDIA_ROW -> FileAndMediaViewHolder(parent)
            else -> EmptyViewHolder(parent)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (getItemViewType(position) == TYPE_DATE_ROW) {
            val dateStampHolder = holder as FilesDateStampHolder
            dateStampHolder.onBind(list[position].timeStamp)
        }
        if (getItemViewType(position) == TYPE_FILE_AND_MEDIA_ROW) {
            val fileAndMediaHolder = holder as FileAndMediaViewHolder
            val (fileDescription) = list[position] as FileAndMediaItem
            fileAndMediaHolder.onBind(
                fileDescription,
                {
                    val (fileItem) = list[fileAndMediaHolder.adapterPosition] as FileAndMediaItem
                    onFileClick.onFileClick(fileItem)
                }
            ) {
                val (fileItem) = list[fileAndMediaHolder.adapterPosition] as FileAndMediaItem
                onDownloadFileClick.onDownloadFileClick(fileItem)
            }
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (list[position] is DateRow) {
            TYPE_DATE_ROW
        } else if (list[position] is FileAndMediaItem) {
            TYPE_FILE_AND_MEDIA_ROW
        } else super.getItemViewType(position)
    }

    /**
     * Очищает текущий список, предварительно создавая его бэкап
     */
    fun backupAndClear() {
        backup = ArrayList(list)
        list.clear()
    }

    /**
     * Фильтрует список по заданному параметру
     * @param filter строка для фильтрации
     */
    fun filter(filter: String?) {
        var filter = filter
        if (filter == null) filter = ""
        list.clear()
        val filteredItems = ArrayList<FileDescription>()
        for (item in backup) {
            if (item is FileAndMediaItem) {
                val (fileDescription, fileName) = item
                val lastPathSegment = fileDescription.fileUri?.lastPathSegment
                val lowercaseLastPath = lastPathSegment?.lowercase(Locale.getDefault())
                val lowercaseFilter = filter.lowercase(Locale.getDefault())
                val isContainsThisLocale = lowercaseLastPath?.contains(lowercaseFilter) == true

                if (isContainsThisLocale) {
                    filteredItems.add(fileDescription)
                } else if (fileDescription.incomingName != null) {
                    val lowercaseName = fileDescription.incomingName.lowercase(Locale.getDefault())
                    if (lowercaseName.contains(lowercaseFilter)) {
                        filteredItems.add(fileDescription)
                    }
                } else if (fileName.lowercase(Locale.getDefault()).contains(lowercaseFilter)) {
                    filteredItems.add(fileDescription)
                }
            }
        }

        updateWithDiffUtil { addItems(filteredItems) }
    }

    /**
     * Восстанавливает список из бэкапа
     */
    fun undoClear() {
        updateWithDiffUtil { list = ArrayList(backup) }
        backup.clear()
    }

    /**
     * Обновляет прогресс загрузки для файла
     * @param fileDescription характеристики файла для обновление прогресса
     */
    fun updateProgress(fileDescription: FileDescription?) {
        for (i in list.indices) {
            if (list[i] is FileAndMediaItem) {
                val mediaItem = list[i] as FileAndMediaItem
                if (ObjectsCompat.equals(mediaItem.fileDescription, fileDescription)) {
                    val itemCopy = mediaItem.copy(
                        fileDescription = fileDescription!!,
                        fileName = mediaItem.fileName
                    )
                    list[i] = itemCopy
                    notifyItemChanged(i)
                }
            }
        }
    }

    /**
     * Показывает ошибку загрузки для файла
     * @param fileDescription характеристики файла для отображения ошибки
     */
    fun onDownloadError(fileDescription: FileDescription?) {
        for (i in list.indices) {
            if (list[i] is FileAndMediaItem) {
                val (fileDescription1) = list[i] as FileAndMediaItem
                val itemViewType = getItemViewType(i)
                if (ObjectsCompat.equals(fileDescription1, fileDescription) &&
                    itemViewType == TYPE_FILE_AND_MEDIA_ROW
                ) {
                    fileDescription1.isDownloadError = true
                    notifyItemChanged(i)
                }
            }
        }
    }

    private fun addItems(fileDescriptionList: List<FileDescription?>) {
        if (fileDescriptionList.isEmpty()) return
        val current = Calendar.getInstance()
        val prev = Calendar.getInstance()
        if (list.isEmpty()) {
            fileDescriptionList[0]?.let { fd ->
                list.add(DateRow(fd.timeStamp))
                list.add(
                    FileAndMediaItem(
                        fd,
                        if (fd.fileUri != null) getFileName(
                            fd.fileUri!!
                        ) else ""
                    )
                )
            }
        }
        for (i in 1 until fileDescriptionList.size) {
            fileDescriptionList[i]?.let { fd ->
                current.timeInMillis = fd.timeStamp
                prev.timeInMillis = fileDescriptionList[i - 1]?.timeStamp ?: 0L
                list.add(
                    FileAndMediaItem(
                        fd,
                        if (fd.fileUri != null) getFileName(
                            fd.fileUri!!
                        ) else ""
                    )
                )
                if (current[Calendar.DAY_OF_YEAR] != prev[Calendar.DAY_OF_YEAR]) {
                    list.add(DateRow(fd.timeStamp))
                }
            }
        }
    }

    private inline fun updateWithDiffUtil(changeItems: () -> Unit) {
        val oldItems = ArrayList(list)
        changeItems()
        val diffCallback = FilesAndMediaDiffUtil(oldItems, list)
        val diff = DiffUtil.calculateDiff(diffCallback)
        diff.dispatchUpdatesTo(this)
    }

    interface OnFileClick {
        /**
         * Описывает реакцию на нажатие
         * @param fileDescription характеристики файла, на котором был произведен клик
         */
        fun onFileClick(fileDescription: FileDescription?)
        /**
         * Описывает реакцию на начало загрузки файла
         * @param fileDescription характеристики файла, на котором был произведен клик
         */
        fun onDownloadFileClick(fileDescription: FileDescription?)
    }

    companion object {
        private const val TYPE_DATE_ROW = 1
        private const val TYPE_FILE_AND_MEDIA_ROW = 2
    }
}
